src/rta/fullsupernodelist.cpp

```diff
@@ -0,0 +1,352 @@
+    }
+    for (int i = 0; i < threads; i++) {
+        m_threadpool.create_thread(boost::bind(&boost::asio::io_service::run, &m_ioservice));
+    }
+}
+
+
+template<typename Callable>
+void ThreadPool::enqueue(Callable job)
+{
+    m_ioservice.dispatch(job);
+}
+
+void ThreadPool::run()
+{
+    m_work.reset();
+    m_work.reset();
+    while (!m_ioservice.stopped())
+        m_ioservice.poll();
+    m_threadpool.join_all();
+    m_ioservice.stop();
+}
+
+std::future<void> ThreadPool::runAsync()
+{
+    return std::async(std::launch::async, [&]() {
+       this->run();
+    });
+}
+
+} // namespace utils;
+
+FullSupernodeList::FullSupernodeList(const string &daemon_address, bool testnet)
+    : m_testnet(testnet)
+    , m_daemon_address(daemon_address)
+    , m_rpc_client(daemon_address, "", "")
+    , m_tp(new utils::ThreadPool())
+{
+    m_refresh_counter = 0;
+}
+
+FullSupernodeList::~FullSupernodeList()
+{
+    boost::unique_lock<boost::shared_mutex> writerLock(m_access);
+    m_list.clear();
+}
+
+bool FullSupernodeList::add(Supernode *item)
+{
+
+    if (exists(item->walletAddress()))
+        return false;
+
+    boost::unique_lock<boost::shared_mutex> writerLock(m_access);
+    m_list.insert(std::make_pair(item->walletAddress(), SupernodePtr{item}));
+    LOG_PRINT_L0("added supernode: " << item->walletAddress());
+    LOG_PRINT_L0("list size: " << m_list.size());
+    return true;
+}
+
+size_t FullSupernodeList::loadFromDir(const string &base_dir)
+{
+    vector<string> wallets = findWallets(base_dir);
+    size_t result = 0;
+    LOG_PRINT_L0("found wallets: " << wallets.size());
+    for (const auto &wallet_path : wallets) {
+        Supernode * sn = Supernode::load(wallet_path, "", m_daemon_address, m_testnet);
+        if (sn)  {
+            if (!this->add(sn)) {
+                LOG_ERROR("Can't add supernode " << sn->walletAddress() << ", already exists");
+                delete sn;
+            } else {
+                LOG_PRINT_L1("Added supernode: " << sn->walletAddress() << ", stake: " << sn->stakeAmount());
+                ++result;
+            }
+        }
+    }
+    return result;
+}
+
+
+
+size_t FullSupernodeList::loadFromDirThreaded(const string &base_dir, size_t &found_wallets)
+{
+    vector<string> wallets = findWallets(base_dir);
+    LOG_PRINT_L0("found wallets: " << wallets.size());
+    found_wallets = wallets.size();
+
+    utils::ThreadPool tp;
+
+    auto worker = [&](const std::string &wallet_path) {
+        Supernode * sn = Supernode::load(wallet_path, "", m_daemon_address, m_testnet);
+        if (sn) {
+            if (!this->add(sn)) {
+                LOG_ERROR("can't add supernode: " << sn->walletAddress());
+                delete sn;
+            } else {
+                LOG_PRINT_L0("Loaded supernode: " << sn->walletAddress());
+            }
+        } else {
+            LOG_ERROR("Error loading supernode wallet from: " << wallet_path);
+        }
+    };
```

Comment:
As for me can be moved to a private method, since it's copied code at Line149

Reply:
fixed with 0f0b66ca84bd6650cec0b9a35319483102053043

Reply:
fixed with 0f0b66ca84bd6650cec0b9a35319483102053043